### PR TITLE
FEAT-#1205: melt implementation

### LINKED ISSUE #1205: melt function implementation

### PR DESCRIPTION
**Pull request by dchigarev**
## What do these changes do?
Implementation of `pandas.DataFrame.melt` + `apply_full_axis_select_indices` optimization.

- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1205 
- [x] tests added and passing

### CODE REVIEW

File: modin/backends/pandas/query_compiler.py [hunk 1]
```diff
@@ -1180,6 +1180,56 @@ def sort_index(self, **kwargs):
...
+                x = [*x]
+            elif x is not None:
+                x = [x]
+            else:
+                x = []
+            return x
+
+        unique_name = "__-index-__"
+
+        # inserting column to sort in a correct order futhure
```
Comment (devin-petersohn) [inline modin/backends/pandas/query_compiler.py:1195]:
```suggestion
        # inserting column to sort in a correct order future
```

File: modin/backends/pandas/query_compiler.py [hunk 2]
```diff
@@ -1180,6 +1180,56 @@ def sort_index(self, **kwargs):
...
+        prepaired = self.insert(
+            loc=len(self.columns), column=unique_name, value=np.arange(len(self.index))
```
Comment (devin-petersohn) [inline modin/backends/pandas/query_compiler.py:1197]: This is going to be slow because you are using a `numpy` array, which is not distributed.

Comment (devin-petersohn) [reply modin/backends/pandas/query_compiler.py:1197]: Is there a way not to use this?

**Comment by dchigarev** [comment]
@devin-petersohn Thanks for your comments, definitely first implementation were extremely slow because of inserting column with `numpy` array and sorting indices via `sort_index`. Now for restoring pandas-order of output it uses [`modin_frame.reorder_labels`](https://github.com/modin-project/modin/blob/master/modin/engines/base/frame/data.py#L450-L483), but it's still slower than pandas:
```
500000x45 ~ 74Mb Dataset

Pandas: 1.0365909719839692
Modin: 4.968821498565376

time of execution reorder_labels inside
modin melt implementation: 4.164067830890417
```

As we see from the results, native modin frame function `reorder_labels` is a bottleneck here.

Here is the results without restoring pandas-order:

```
500000x45 ~ 74Mb Dataset

Pandas: 1.0422626351937652
Modin: 0.5023592244833708
```
<details><summary>How it was measured</summary>


```python
import numpy as np
import string
import csv
import os
from timeit import default_timer as timer

IS_PANDAS = True
RAND_LOW = -100
RAND_HIGH = 100
N = 500000
M = 45
TEST_FILENAME = os.path.abspath(f"int_dataset-{N},{M},{RAND_LOW},{RAND_HIGH}.csv")


def generate_data_file(filename, row_n, col_n):
    letters = list(string.ascii_letters)
    columns = [
        "".join(np.random.choice(letters) for _ in np.arange(10))
        for _ in np.arange(col_n)
    ]

    with open(filename, "w") as file:
        writer = csv.writer(file, delimiter=",")
        writer.writerow(columns)
        for _ in np.arange(row_n):
            writer.writerow(np.random.randint(RAND_LOW, RAND_HIGH, col_n))


if __name__ == "__main__":
    import pandas
    import modin.pandas as pd

    if not os.path.exists(TEST_FILENAME):
        generate_data_file(TEST_FILENAME, N, M)

    if IS_PANDAS:
        pd_df = pandas.read_csv(TEST_FILENAME)
        t1_pd = timer()
        pd_df.melt()
        t2_pd = timer()

        print("Pandas:", t2_pd - t1_pd)

    md_df = pd.read_csv(TEST_FILENAME)
    t1_md = timer()
    md_df.melt()
    t2_md = timer()

    print("Modin:", t2_md - t1_md)

```


</details>

<details><summary>What is restoring pandas-order?</summary>



For applying `melt` function we must get information about full row axis, so it's implemented via `modin_frame._apply_full_axis(1, lambda df: df.melt())` and because function applied only for row partitions, we would get a different order of rows as the result.

For example we have DataFrame splitter into row partitions like that:
```
   first     last  weight  height
0   John    Smith     5.0     130
1   Mary  Johnson     6.0     140
----------------------------------
2    Bob    Jones     5.5     150
3  Andry    Brown     4.2     160
```

and then we will apply `melt(id_vars=["first", "last"])`

```
Result of applying			Result of applying
for row partitions:			for the whole dataframe:

   first     last variable  value	   first     last variable  value	
0   John    Smith   weight    5.0	0   John    Smith   weight    5.0
1   Mary  Johnson   weight    6.0	1   Mary  Johnson   weight    6.0
2   John    Smith   height  130.0	2    Bob    Jones   weight    5.5
3   Mary  Johnson   height  140.0	3  Andry    Brown   weight    4.2
---------------------------------
0    Bob    Jones   weight    5.5	4   John    Smith   height  130.0
1  Andry    Brown   weight    4.2	5   Mary  Johnson   height  140.0
2    Bob    Jones   height  150.0	6    Bob    Jones   height  150.0
3  Andry    Brown   height  160.0	7  Andry    Brown   height  160.0
```
Restoring order from right result in left result is what I mean under **restoring pandas-order**

</details>

Restoring order is the most expensive operation because we must have knowledge about the whole column axis, which is the longest one in case of `melt` output, so the reordering will take longer that `melt` itself.

So, @devin-petersohn is it acceptable for modin to output shuffled data in such functions or modin has a pattern for fast reordering which I'm don't know?

**Comment by anmyachev** [comment]
@dchigarev maybe you can found some useful things in #725

File: modin/backends/pandas/query_compiler.py [hunk 3]
```diff
@@ -1180,6 +1180,52 @@ def sort_index(self, **kwargs):
...
+        def _convert_to_list(x):
+            if is_list_like(x):
+                x = [*x]
+            elif x is not None:
+                x = [x]
+            else:
+                x = []
+            return x
+
+        row_lengths = self._modin_frame._row_lengths
```
Comment (devin-petersohn) [inline modin/backends/pandas/query_compiler.py:1193]: Let's try to figure out a way to avoid accessing and relying on this field

File: modin/backends/pandas/query_compiler.py [hunk 4]
```diff
@@ -1180,6 +1180,52 @@ def sort_index(self, **kwargs):
...
+        id_vars, values_vars = map(
+            _convert_to_list,
+            [kwargs.get("id_vars", None), kwargs.get("value_vars", None)],
+        )
+
+        value_vars = len(values_vars)
+        if value_vars == 0:
+            value_vars = len(self.columns) - len(id_vars)
+
+        shuffled = self._modin_frame._apply_full_axis(
```
Comment (devin-petersohn) [inline modin/backends/pandas/query_compiler.py:1204]: I understand why you chose `_apply_full_axis` at first. Do you think it is possible to use `_apply_select_indices`? We will need to potentially replicate the untouched partitions. It will also improve performance significantly.

Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:1204]: It could be implemented via `_apply_select_indices` but then we will have to broadcast partitions of `id_vars` columns to every partition, which is not so fast since we have multicolumn `id_vars`, for example if `modin_frame` with `id_vars` columns has more than one column partitions we can't just do that shiny broadcast like we do in [`groupby_reduce`](https://github.com/modin-project/modin/blob/master/modin/engines/base/frame/partition_manager.py#L78-L97), in that case we gotta use approach from [`binary_operation`](https://github.com/modin-project/modin/blob/master/modin/engines/base/frame/partition_manager.py#L635-L672) which is actually applying function at the full axis.

Nevertheless, even if we could implement that with `_apply_select_indices` that will not solve the problem with slow reordering after applying `melt`. Any thoughts about reordering?

Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py]: Since `melt` is made for wide data frames, I'll try to reimplement it with `_apply_full_axis_select_indices(0, ...)`. If initially we have wide DataFrame with a lot of column partitions, so it should scale pretty well if we'll apply function at columns axis.

File: modin/backends/pandas/query_compiler.py (hunk 2 above)
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py]: `melt` reimplemented

File: modin/backends/pandas/query_compiler.py (hunk 3 above)
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py]: reimplemented

**Comment by dchigarev** [comment]
@devin-petersohn I've reimplemented `melt` via `apply_full_axis_select_indices` and got this results
```
45x500000 (5 values in a different partitions)
Pandas: 0.1547
Modin (8 cores): 0.1028
Modin (112 cores): 0.0514
Modin default_to_pandas: 2.0192

45x500000 (all values)
Pandas: 0.4206
Modin (8 cores): 0.9479
Modin (112 cores): 0.8355
Modin default_to_pandas: 3.3246
```
I've also fixed a strange bug in [`apply_func_to_select_indices_along_full_axis`](https://github.com/modin-project/modin/pull/1689/commits/1b72da2f3c67eb56758b79c9a70ac95511387b9e), without this fix I was getting the same times when applying `melt` function for only one column partition and for the whole column partitions. Does `cls.column_partitions()` do any hard work if we don't apply a function to that column partition? If yes, then for me it's not obvious and I guess it should be mentioned in documentation.

File: modin/engines/base/frame/partition_manager.py [hunk 5]
```diff
@@ -526,11 +526,19 @@ def apply_func_to_select_indices_along_full_axis(
         # Since we might be keeping the remaining blocks that are not modified,
         # we have to also keep the block_partitions object in the correct
         # direction (transpose for columns).
+        if not keep_remaining:
```
Comment (devin-petersohn) [inline modin/engines/base/frame/partition_manager.py:529]: Why was this change necessary? The condition here is handled below.

File: modin/backends/pandas/query_compiler.py [hunk 6]
```diff
@@ -1385,6 +1385,75 @@ def sort_index(self, **kwargs):
...
+                x = []
+            return x
+
+        id_vars, value_vars = map(_convert_to_list, [id_vars, value_vars])
+
+        if len(value_vars) == 0:
+            value_vars = self.columns.drop(id_vars)
+
+        if len(id_vars) != 0:
+            id_vars_df = self.getitem_column_array(id_vars).to_pandas()
```
Comment (devin-petersohn) [inline modin/backends/pandas/query_compiler.py:1414]: If we are materializing the object, we should probably give a warning about defaulting to pandas implementation because this part cannot be done in a distributed way.

**Comment by dchigarev** [comment]
> @dchigarev Yes, if there is no work being done, the column partition is still shuffled together to the same worker, which is some significant work.
> 
> Thanks for running the performance numbers on this. Is there a reason this cannot be done by applying the function to row partitions instead of column partitions?

@devin-petersohn It could be implemented via applying to row partitions, but there is a two details:

1. `melt` is designed to transform dataframe from **wide to long** format, since we have wide dataframe with short columns it will be more scalable if we'll apply function to column partitions, not to rows
2. Applying `melt` to rows leads to losing correct rows order in the result, that problem was described in [that comment](https://github.com/modin-project/modin/pull/1689#issuecomment-657705353). So we will have to do slow reordering. Unfortunately, I still have no idea how to do it fast with the current `modin_frame` API

**Comment by devin-petersohn** [comment]
> Applying melt to rows leads to losing correct rows order in the result, that problem was described in that comment. So we will have to do slow reordering. Unfortunately, I still have no idea how to do it fast with the current modin_frame API

Ah, I forgot about that, sorry. The reordering rows can happen via sort, which isn't implemented scalably yet, but should be soon. It is okay to rely on functionality that isn't implemented yet. We could also perform the sort first, which would probably perform better. What do you think about this?

File: modin/engines/base/frame/partition_manager.py (hunk 5 above)
Comment (dchigarev) [reply modin/engines/base/frame/partition_manager.py:529]: It's not exactly the same condition, as you said above:
>the column partition is still shuffled together to the same worker, which is some significant work

so that changes are to do not creating column partitions for that partitions which will not be considered at computations.
Also, I've got these results after/before this changes:
```
45x500000, 112 cores

With fix:
Modin (1 value): 0.0579
Modin (all values): 0.9072

Without fix:
Modin (1 value):  0.2299
Modin (all values): 0.8566
```

**Comment by dchigarev** [comment]
@devin-petersohn what do you mean under "perform the sort first"? The problem is that after applying `melt` rows are sorted correctly, but only for every independent partition, so it requires some kind of merge phase to sort it globally. Also as you can see in the example above, rows sorted not in alphabetic order, but in the order when they were columns, will it be possible to pass a custom order function with the new implemented `sort`?

And another question, I don't know how data scientists use `melt`, but in my suggestions, it is transforming wide and short dataframes to a long format (correct me if I'm wrong), so in that case, applying to columns will be more efficient?

File: modin/backends/pandas/query_compiler.py (hunk 6 above)
Comment (dchigarev) [reply modin/backends/pandas/query_compiler.py:1417]: Added a warning message

**Comment by devin-petersohn** [comment]
@dchigarev It will scale better if operations are performed across the columns `axis=1`. We can throw a `Warning` that mentions that the order may not match pandas exactly for a first pass so that we do not pay the heavy penalty for reordering the rows. Please create an issue for the ordering of the rows in `melt` implementation so we can fix it later.

**Comment by dchigarev** [comment]
@devin-petersohn I've reimplemented `melt` with applying function to selected indices + broadcasting columns from `id_vars` to every partition. This is the results:
```
500000x128 (all values)
Pandas: 2.9566
Modin (8 cores): 0.6220
Modin (112 cores): 0.5468

500000x128 (5 values in a different partitions)
Pandas: 0.1019
Modin (8 cores): 0.0893
Modin (112 cores): 0.3803 (re-runed many times, but still getting 
                                 that weird result for 112 cores)
```

Also added a warning of incorrect order of rows and created an issue, to we could fix this later (#1793)

**Comment by devin-petersohn** [comment]
Thanks @dchigarev! Were those times including a print statement to ensure computation was finished?

**Comment by dchigarev** [comment]
@devin-petersohn It's shamable, but it seems that I forgot to add `repr` to the result to force materialization in my measurement script, so all measurements above are not relevant. Time measurements with `repr` are not so magical:
```
Pandas: 2.9670
Modin (8 cores): 1.5371
Modin (112 cores): 1.1135
```

**Comment by devin-petersohn** [comment]
@dchigarev 500,000 rows is not very many, so that speedup is quite good, especially on 128 columns.

**Comment by devin-petersohn** [comment]
@dchigarev The implementation looks good, I have retriggered GitHub Actions because it showed a -40% coverage drop.

**Comment by dchigarev** [comment]
@devin-petersohn now codecov check is okay, I guess
